Update Product Details and Product Reviews to new version

dinhtungdu opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR:
- updates the Product Details and Product Reviews block to the "blockified" version
- maintains the backward compatibility for stores that are using those blocks by conditionally rendering the previous version.
- uses deprecation API to handle the migration in the Editor.


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:
1. On trunk, add the Product Details and Product Reviews to the Single Product template.
2. Visit a product on the frontend, verify blocks added in step 1 are rendered normally.
1. Checkout and build this PR using production build config.
3. Visit a product on the frontend, verify blocks added in step 1 are rendered as step 2.
4. Edit the Single Product template.
5. Verify the two blocks are migrated to new versions (with inner blocks). Verify that there is no block validation error.
6. Make change and update the template.
7. Verify the new version of those blocks  are rendered on the frontend.
8. Reset the Single Product template.
9. Verify you can search and add the `Product Details` block to the page and it's the "blockified" version.
4. Verify you can search and add the `Product Reviews` block to the page and it's the "blockified" version.


<!-- End testing instructions -->

dinhtungdu committed:
fix: unit test

dinhtungdu committed:
fix: e2e test

dinhtungdu committed:
feat: merge two product reviews block

dinhtungdu committed:
feat: merge two product details blocks

dinhtungdu committed:
chore: changelog

dinhtungdu committed:
chore: update template

dinhtungdu committed:
chore: update product-details block name usage

dinhtungdu committed:
chore: update product-reviews block name usage

dinhtungdu committed:
test: update unit test class names

dinhtungdu committed:
chore: remove product details legacy files

dinhtungdu committed:
chore: remove product reviews legacy files

dinhtungdu commented on plugins/woocommerce/client/blocks/assets/js/atomic/blocks/product-elements/product-details/style.scss:
Can we remove this style? Is that okay to drop support for users that are currently using this style.

gigitux replied:
I noticed two things. The style variation selector isn't present anymore. It is not caused by your PR, but it is a regression present since WooCommerce 9.5.0 😞 . More context: p1750426737141959-slack-C02SGH7JBGS

However, removing this style creates a different layout for the stores that have the `minimal` style variation enabled:

| Before | After |
|--------|--------|
|<img width="1934" alt="image" src="https://github.com/user-attachments/assets/41df8076-4fd8-403e-94aa-004fe46aa3bd" /> | <img width="1907" alt="image" src="https://github.com/user-attachments/assets/40553b9e-c22b-4b2d-b671-4bf1fc35b4ab" /> |

dinhtungdu replied:
I added the style back in [b0929e0](https://github.com/woocommerce/woocommerce/pull/59005/commits/b0929e024803c7b975991b2278499d3a2518f0a7)

dinhtungdu commented on plugins/woocommerce/client/blocks/assets/js/atomic/blocks/product-elements/product-reviews/style.scss:
Same question, can we delete this one?

gigitux replied:
I think yes. This style is necessary only for the editor preview: https://github.com/woocommerce/woocommerce-blocks/pull/8857

gigitux commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/index.tsx:
```diff
@@ -14,6 +14,13 @@ const blockConfig = {
+	],
 };
 // @ts-expect-error metadata is not typed.
 registerProductBlockType( blockConfig, {
```
Currently, the layout of the block is broken when it is used in the post editor within the Single Product Block:

<img width="1876" alt="image" src="https://github.com/user-attachments/assets/26e41599-8223-4104-9c3a-4bb9cc04bf73" />

I would update `isAvailableOnPostEditor` to false.

dinhtungdu replied:
I can't reproduce this, the block style loads as expected for me. Can you check it again?
<img width="1025" alt="image" src="https://github.com/user-attachments/assets/27ab88fb-96d8-4050-982f-b57ac84a528c" />

gigitux replied:
Yes! I did another test and these are my findings, adding the Product Reviews in a post.

### Legacy Mode - no inner blocks

| Without comments | With comments |
|--------|--------|
| <img width="1513" alt="image" src="https://github.com/user-attachments/assets/13ec20b5-de8a-46bd-b158-5537fb0086c5" /> | <img width="1589" alt="image" src="https://github.com/user-attachments/assets/a50953d9-30d5-4f0e-ab7e-62bc3edcf498" /> | 

It looks like that the missing some style.

### With inner blocks

| Without comments | With comments |
|--------|--------|
| <img width="1167" alt="image" src="https://github.com/user-attachments/assets/b2dcfbbc-d765-4415-bf7a-ca6aa61af465" /> | ![image](https://github.com/user-attachments/assets/83c0d099-34b5-455c-beb3-97b08dc2d299) | 

I think that the select should be replaced with stars, right?

dinhtungdu replied:
Thanks for additional testing, there are so many issues using these blocks inside Single Product. Let's first disable it. We can revisit this usage case latter.

dinhtungdu replied:
Addressed in https://github.com/woocommerce/woocommerce/pull/59005/commits/fa14d4a9607cffcb4b0b6b74469764abb2a0ca73

gigitux commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-details/index.tsx:
```diff
@@ -16,6 +16,23 @@ const blockConfig = {
+	],
 };
 // @ts-expect-error blockConfig is not typed.
 registerProductBlockType( blockConfig, {
```
We are not registering this block on the[ post editor in the PHP side](https://github.com/woocommerce/woocommerce/blob/3fda7251cedfbece5f66ae830d84285cfcaa8e23/plugins/woocommerce/src/Blocks/BlockTypesController.php#L579-L580). It would make sense to update the flag here

dinhtungdu replied:
Do you know why do we do it in the first place?

gigitux replied:
I'm not sure, but I think that the PHP wrapper block (the old Product Details) doesn't work fine in the post/page editor context

dinhtungdu replied:
You're right, if added to the single post or page, that block will break the site because of `rewind_posts()` used in its render callback.

dinhtungdu replied:
Covered in https://github.com/woocommerce/woocommerce/pull/59005/commits/2efd4b589b0a64b63d415e728cc49e7aaa6d8acb

gigitux commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-details/block.json:
Should we update the `ancestor` property?

dinhtungdu replied:
What should be the ideal value? Do you mean removing the whole property?

gigitux replied:
Is it expected that the block can be added to the Product Collection block? I would make it available only in the Single Product. At the same time, with fresh mind, why we should add a constraint if the block works fine? 

So, now, I'm not sure! What do you think?

dinhtungdu replied:
> Is it expected that the block can be added to the Product Collection block? I would make it available only in the Single Product. 

Hmm, no, I don't think the block should be added inside loop. I don't have the context why it's added but I think we should remove "woocommerce/product-template" and "core/post-template".

dinhtungdu replied:
> At the same time, with fresh mind, why we should add a constraint if the block works fine?

We haven't tested it in PC yet. It probably works, but we're not sure. So it's better to restrict.

gigitux replied:
> We haven't tested it in PC yet. It probably works, but we're not sure. So it's better to restrict.

It makes sense 👍

dinhtungdu replied:
I ended up remove the whole ancestor property because we disable the block in the site editor: https://github.com/woocommerce/woocommerce/pull/59005/commits/fa14d4a9607cffcb4b0b6b74469764abb2a0ca73

gigitux commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/block.json:
Should we update the `ancestor` property?

dinhtungdu replied:
Merged with the same comment on the product details.

gigitux commented on plugins/woocommerce/templates/templates/blockified/single-product.html:
As you mentioned, we should not update this template.

gigitux commented on plugins/woocommerce/src/Blocks/BlockTypes/Reviews/ProductReviews.php:
```diff
@@ -39,4 +44,33 @@ protected function render( $attributes, $content, $block ) {
+	 *
+	 * @return string Rendered block output.
+	 */
+	protected function render_legacy_block( $attributes, $content, $block ) {
```
Should we write more details? In the long term and for new developers could be complex to understand why this is necessary.

gigitux replied:
I would write a unit test to ensure that the `render_legacy_block` is invoked when `innerBlocks` is empty. Things can evolve, and we can easily forget to test this scenario.

dinhtungdu replied:
Makes sense! I addressed this in [b9070b1](https://github.com/woocommerce/woocommerce/pull/59005/commits/b9070b153d49278081d8f9702b4c6db279c08b9a).

gigitux reviewed (changes requested):
Thanks @dinhtungdu for tackling this. Great work! I’ve left a few comments, but my main concern is about the missing style variations for the Product Details block.

Currently, the style variations don’t appear to be registered. If the goal is to drop support for these variations (since they haven’t worked since WooCommerce 9.5.0), this approach makes sense. Otherwise, we’d need to create separate blocks.

Personally, I think it’s fine to remove them, especially since no one has reported the issue. The only important thing is to keep `blocks/product-elements/product-details/style.scss`, as it’s required to load the default `minimal` style variation.

gigitux commented:
Testing this PR, I found a bug in Gutenberg. During the drag and drop, in some cases, the inner blocks are re-added. This bug happens with a Core Block, so it is not related to your PR. 

| Woo | Core |
|--------|--------|
| <video src="https://github.com/user-attachments/assets/e56ee769-dfff-4bdf-9efa-425f5ceb01f8" /> |<video src="https://github.com/user-attachments/assets/d0f5c1bd-ac15-4f2a-aa7a-d630356aca82" /> |

dinhtungdu committed:
Revert "chore: update template"

This reverts commit 25acc66ea4a834e6bad33921e0081be12a410809.

dinhtungdu committed:
fix: backward compatible minimal style for product details

dinhtungdu commented:
> Do you think this is a bug or expected behavior?

I think it's both, it's expected for block to use the inner block template when created. It looks like a bug that the block is sometime recreated while moving it.

dinhtungdu committed:
docs: explain why we need the legacy render method

dinhtungdu committed:
test: legacy block render

dinhtungdu commented:
<!-- linear:isThreadRoot -->

> I think that the block, when used in the legacy mode, crashes the WordPress site when it is rendered on the homepage:

@gigitux Yup, it's because of `rewind_posts()` used in its render callback. I will add some defensive checks to prevent it from happening.

dinhtungdu committed:
fix: disable product details and reviews from the post editor due to fatal error with rewind_posts().

Because we disable it in the editor, it doesn't make sense to define block ancestor anymore.

dinhtungdu committed:
fix: ensure the block doesn't break the site when actidentally used outside single product template

dinhtungdu committed:
docs: explain the legacy render method of product reviews

dinhtungdu committed:
Revert "test: legacy block render"

This reverts commit 37f406b9b208a2b48038a29902ce6d1aa1cc1245.

dinhtungdu commented:
Reverting the new unit test case in b6b25d6 because we added a defensive check to only render the block on the single product page ([2efd4b5](https://github.com/woocommerce/woocommerce/pull/59005/commits/2efd4b589b0a64b63d415e728cc49e7aaa6d8acb). We need to meet 10.0 for this PR so skipping unit test for now.

gigitux reviewed (approved):
I can confirm the following behaviors:

- **By default**, the legacy render function is invoked:  
  <img width="1786" alt="image" src="https://github.com/user-attachments/assets/470209ee-5562-4093-ba4b-23c1e00c1e0b" />

- **When the save function is triggered**, the new render function is used:  
  <img width="1795" alt="image" src="https://github.com/user-attachments/assets/50405295-cb05-40a2-b256-74eb6ded238d" />

- **If the Product Details block has no inner blocks**, the legacy render function is used again:  
  <img width="1821" alt="image" src="https://github.com/user-attachments/assets/6dfee059-d304-4fad-8079-f7a9ab8e079b" />


LGTM!

gigitux commented:
> I think it's both; it's expected for the block to use the inner block template when created. It looks like a bug that the block is sometimes recreated while moving it. I still think it's worth opening an issue if it doesn't exist.

Thanks for the feedback! I will open an issue soon! 🙇

dinhtungdu committed:
fix: remove the mock because we're registering block for any themes

Previously, we only register the block for block themes. But now, we
upgde the existing blocks, which is registered for any themes. So we
don't need to manually initialize the block class anymore.

dinhtungdu committed:
Merge branch 'release/10.0' into fix/remove-product-details-feature-gate

dinhtungdu commented:
> Cherry picking failed for the trunk.

New PR is manually created for trunk: https://github.com/woocommerce/woocommerce/pull/59105
